src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -1463,19 +1435,58 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
      * @param predicate the predicate to evaluate.
      * @return true if succesful.
      */
-    private boolean tryTransferFromPosToWorker(final BlockPos pos, @NotNull final Predicate<ItemStack> predicate)
+
+    private boolean tryTransferFromPosToWorker(final BlockPos pos, @NotNull final Tuple<Predicate<ItemStack>, Integer> predicate)
     {
         final TileEntity entity = world.getTileEntity(pos);
         if (entity == null)
         {
             return false;
         }
 
-        return InventoryUtils.transferXOfFirstSlotInProviderWithIntoNextFreeSlotInItemHandler(
-          entity,
-          predicate,
-          Constants.STACKSIZE,
-          worker.getInventoryCitizen());
+        final int existingAmount = InventoryUtils.getItemCountInItemHandler(worker.getInventoryCitizen(), predicate.getA());
+        int amount = 0;
+        if (predicate.getB() > existingAmount)
+        {
+            amount = predicate.getB() - existingAmount;
+        }
+        else
+        {
+            return true; // has already needed transfers...
```

Comment:
sure this is right? tryTransferFromPosToWorker doesnt seem like a function which checks the own inventory for contains

Reply:
this is why i added getItemCountInItemHandler. which assists in picking the correct amount. like if i need 2 stacks he will pick 2 stacks ( the Do while loop ). prevents that a Builder walks multiple times for the same stuff.

Reply:
ye then rename the function to sth like transferItemsIfNeeded or so

Reply:
i can. is only used by pickupMaterials() anyway.